### ISSUE TITLE
BUG: pchip interpolation of complex values is buggy due to sign definition

### ISSUE THREAD

**Issue by mhvk** [scipy/scipy#19739]
b### Describe your issue.

The definition of `pchip` interpolation relies on noticing zeros and sign changes in the derivative. This works for real numbers but not for complex -- if complex is desired, one should interpolate real and imaginary separately.

I should add that the docstring explicitly states that `y` should be real, but there is a test with complex numbers anyway -- `test_rgi.py::TestRegularGridInterpolator::test_complex[pchip]`. And this test starts failing with a PR in numpy that changes the definition of the complex sign to `z / |z|` (see https://github.com/numpy/numpy/pull/25441#issuecomment-1867887200). 

This failure is not directly related to the change, though: the present test happens to pass because the values are just a linearly increasing list, and because the real and imaginary are the same modulo a factor. Even with the present definition the test can easily be made to fail (see below).

### Reproducing Code Example

```python
# Change the test case as follows,
$ diff test_rgi.py~ test_rgi.py
46c46
<         values = np.asarray([0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0])
---
>         values = np.asarray([0.0, 0.5, 1.0, 1.5, 2.5, 2.0, 3.0])
123c123
<         values = values - 2j*values
---
>         values = values - 2j*values[::-1]

# Or, as a copy-and-pastable example,
import numpy as np
from scipy.interpolate import RegularGridInterpolator

points = [(0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0)] * 4
values = np.asarray([0.0, 0.5, 1.0, 1.5, 2.5, 2.0, 3.0])
values0 = values[:, np.newaxis, np.newaxis, np.newaxis]
values1 = values[np.newaxis, :, np.newaxis, np.newaxis]
values2 = values[np.newaxis, np.newaxis, :, np.newaxis]
values3 = values[np.newaxis, np.newaxis, np.newaxis, :]
values = (values0 + values1 * 10 + values2 * 100 + values3 * 1000)
values = values - 2j*values[::-1]
sample = np.asarray([[0.1, 0.1, 1., .9], [0.2, 0.1, .45, .8],
                     [0.5, 0.5, .5, .5]])
method = 'pchip'
interp = RegularGridInterpolator(points, values, method=method)
rinterp = RegularGridInterpolator(points, values.real, method=method)
iinterp = RegularGridInterpolator(points, values.imag, method=method)

v1 = interp(sample)
v2 = rinterp(sample) + 1j*iinterp(sample)
print(np.abs(v1 - v2))
```

### Error message

```shell
# Difference is not zero:
[0.09121403 0.2736421  0.        ]
```

### SciPy/NumPy/Python version and system information

```shell
1.11.4 1.26.2 sys.version_info(major=3, minor=11, micro=7, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.11.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.36
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-lzd6dam0/overlay/lib/python3.11/site-packages/pythran
    version: 0.14.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: true
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp311-cp311/bin/python
  version: '3.11'
```

**Comment by lucascolley** [comment]
Hmm yeah, the docstring of `scipy.interpolate.RegularGridInterpolator` says "Complex data can be acceptable" which isn't great wording.

Looks like we need to split the methods between the following two tests depending on whether they are supposed to accept complex input or not (and update the docstring of RGI). So it seems like `pchip` should raise a `ComplexWarning` too?

```python
    @parametrize_rgi_interp_methods
    def test_complex(self, method):
        x, y, values = self._sample_2d_data()
        points = (x, y)
        values = values - 2j*values

        sample = np.array([[1, 2.3, 5.3, 0.5, 3.3, 1.2, 3],
                           [1, 3.3, 1.2, 4.0, 5.0, 1.0, 3]]).T

        v1 = interpn(points, values, sample, method=method)
        v2r = interpn(points, values.real, sample, method=method)
        v2i = interpn(points, values.imag, sample, method=method)
        v2 = v2r + 1j*v2i
        assert_allclose(v1, v2)

    def test_complex_spline2fd(self):
        # Complex-valued data not supported by spline2fd
        x, y, values = self._sample_2d_data()
        points = (x, y)
        values = values - 2j*values

        sample = np.array([[1, 2.3, 5.3, 0.5, 3.3, 1.2, 3],
                           [1, 3.3, 1.2, 4.0, 5.0, 1.0, 3]]).T
        with assert_warns(ComplexWarning):
            interpn(points, values, sample, method='splinef2d')
```

**Comment by mhvk** [comment]
Indeed, `pchip` should also warn (with perhaps an issue raised to adjust it to interpolate real and complex separately, though interpolation of complex numbers is a bit less well defined anyway).

**Comment by ev-br** [comment]
tl;dr: the test is wrong, `pchip` mode should be removed from this test.

Basically, pchip construction is a non-linear in data, so just blindly using the pchip prescription (https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.PchipInterpolator.html) on complex-valued data is not the same as using real and complex parts separately. 

Demo:

```
In [1]: import numpy as np

In [2]: from scipy.interpolate import pchip

In [3]: x = np.arange(8)

In [4]: y = np.sin(x) + 1j*np.cos(x)

In [5]: x2 = (x[1:] + x[:-1]) /2.

In [6]: pchip(x, y.real)(x2)
Out[6]: 
array([ 0.55858037,  0.89107599,  0.62870848, -0.37009501, -0.89910938,
       -0.7176118 ,  0.15412137])

In [7]: pchip(x, y)(x2).real
Out[7]: 
array([ 0.50048003,  0.94917633,  0.66041756, -0.3537783 , -0.94713517,
       -0.75030574,  0.18681531])
```

`Akima1DInterpolator` has the same issue.

Ways towards sanity:

- Remove `pchip` from that RGI test.
- Decide what to do with pchip and Akima1D when `y.dtype` is complex. Options: 
     1. just raise because it returns nonsense. 
     2. redefine pchip and akima for complex `y` to interpolate real and imaginary parts separately. This has no mathematical meaning and breaks back compat just the same.
     3. Improve the docs and leave it be.

I'd go for 3. because it's simplest and has minimal user impact.

Removing `pchip` from the RGI test is a straighforward uncontroversial fix, can be done in a quick PR, and we can even backport it if we want to improve compatibility of SciPy 1.12 with the upcoming NumPy 2.0.

**Comment by mhvk** [comment]
Yes, I was wondering whether interpolating in real and imag separately made sense for non-linear... Given that it doesn't, I'd suggest to raise an exception - people may be getting hit with bad results, better that they know!

**Comment by lucascolley** [comment]
> Decide what to do with pchip and Akima1D when y.dtype is complex. Options:
> 1. just raise because it returns nonsense.
> 2. redefine pchip and akima for complex y to interpolate real and imaginary parts separately. This has no mathematical meaning and breaks back compat just the same.
> 3. Improve the docs and leave it be.

We have +1 for option 1 and +1 for option 3. @j-bowhay option 1 would require a deprecation, right? Any preference? I'm happy to submit a PR for either option.

**Comment by rgommers** [comment]
Deprecating and then removing support for complex inputs seems like the way to go here.

**Comment by j-bowhay** [comment]
> > Decide what to do with pchip and Akima1D when y.dtype is complex. Options:
> > 
> > 1. just raise because it returns nonsense.
> > 2. redefine pchip and akima for complex y to interpolate real and imaginary parts separately. This has no mathematical meaning and breaks back compat just the same.
> > 3. Improve the docs and leave it be.
> 
> We have +1 for option 1 and +1 for option 3. @j-bowhay option 1 would require a deprecation, right? Any preference? I'm happy to submit a PR for either option.

+1 for option 1 too
